File: pkgs/npm-genotype-types/src/core/meta.ts
import { z } from "zod";

export const GtMetaPathsLang = z.object({
  pkg: z.string(),
  src: z.string(),
});

export type GtMetaPathsLang = z.infer<typeof GtMetaPathsLang>;

export const GtMetaPaths = z.object({
  src: z.string(),
  dist: z.string(),
  ts: z.union([GtMetaPathsLang, z.undefined()]).optional(),
  rs: z.union([GtMetaPathsLang, z.undefined()]).optional(),
  py: z.union([GtMetaPathsLang, z.undefined()]).optional(),
});

export type GtMetaPaths = z.infer<typeof GtMetaPaths>;

export const GtMetaTiming = z.object({
  totalMs: z.number(),
  loadProjectMs: z.number(),
  loadModulesMs: z.number(),
  compileMs: z.number(),
});

export type GtMetaTiming = z.infer<typeof GtMetaTiming>;

export const GtMetaModule = z.object({
  source: z.string(),
  ts: z.union([z.string(), z.undefined()]).optional(),
  rs: z.union([z.string(), z.undefined()]).optional(),
  py: z.union([z.string(), z.undefined()]).optional(),
});

export type GtMetaModule = z.infer<typeof GtMetaModule>;

export const GtMeta = z.object({
  exitCode: z.number(),
  paths: GtMetaPaths,
  timing: GtMetaTiming,
  modules: z.array(GtMetaModule),
});

export type GtMeta = z.infer<typeof GtMeta>;

File: pkgs/crate-genotype-core/src/types/meta.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct GtMeta {
    #[serde(rename = "exitCode")]
    pub exit_code: i32,
    pub paths: GtMetaPaths,
    pub timing: GtMetaTiming,
    pub modules: Vec<GtMetaModule>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct GtMetaPaths {
    pub src: String,
    pub dist: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub ts: Option<GtMetaPathsLang>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub rs: Option<GtMetaPathsLang>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub py: Option<GtMetaPathsLang>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct GtMetaPathsLang {
    pub pkg: String,
    pub src: String,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct GtMetaTiming {
    #[serde(rename = "totalMs")]
    pub total_ms: u64,
    #[serde(rename = "loadProjectMs")]
    pub load_project_ms: u64,
    #[serde(rename = "loadModulesMs")]
    pub load_modules_ms: u64,
    #[serde(rename = "compileMs")]
    pub compile_ms: u64,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct GtMetaModule {
    pub source: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub ts: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub rs: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub py: Option<String>,
}
